R/RLearner_classif_mxff.R

```diff
@@ -0,0 +1,416 @@
+    }
+
+    # construct hidden layers using symbols
+    for (i in seq_len(layers)) {
+      if (convs[i]) {
+        # construct convolutional layer with pooling
+        conv.inputs = list(data = sym, kernel = conv.kernels[[i]], stride = conv.strides[[i]],
+          dilate = conv.dilates[[i]], pad = conv.pads[[i]], num_filter = nums[i])
+        sym = do.call(mxnet::mx.symbol.Convolution, conv.inputs[!sapply(conv.inputs, is.null)])
+        sym = mxnet::mx.symbol.Activation(sym, act_type = act[i])
+        pool.inputs = list(data = sym, kernel = pool.kernels[[i]], pool.type = pool.types[[i]],
+          stride = pool.strides[[i]], pad = pool.pads[[i]])
+        sym = do.call(mxnet::mx.symbol.Pooling, pool.inputs[!sapply(pool.inputs, is.null)])
+      } else {
+        # construct fully connected layer
+        if (i > 1) {
+          if (convs[i - 1]) {
+            sym = mxnet::mx.symbol.flatten(sym)
+          }
+        }
+        sym = mxnet::mx.symbol.FullyConnected(sym, num_hidden = nums[i])
+        sym = mxnet::mx.symbol.Activation(sym, act_type = act[i])
+      }
+      # add dropout if specified
+      if (!is.null(dropout)) {
+        sym = mxnet::mx.symbol.Dropout(sym, p = dropout[i + 1])
+      }
+    }
+
+    # construct output layer
+    if (convs[layers]) {
+      sym = mxnet::mx.symbol.flatten(sym)
+    }
+    nodes.out = switch(act.out,
+      softmax = nlevels(d$target),
+      logistic = 1,
+      stop("Output activation not supported yet."))
+    sym = mxnet::mx.symbol.FullyConnected(sym, num_hidden = nodes.out)
+    out = switch(act.out,
+      # rmse = mxnet::mx.symbol.LinearRegressionOutput(sym),
+      softmax = mxnet::mx.symbol.SoftmaxOutput(sym),
+      logistic = mxnet::mx.symbol.LogisticRegressionOutput(sym),
+      stop("Output activation not supported yet."))
+  }
+
+  # create model
+  model = mxnet::mx.model.FeedForward.create(out, X = x, y = y, eval.data = eval.data,
+    epoch.end.callback = epoch.end.callback, array.layout = array.layout, ...)
+  return(model)
+}
+
+#' @export
+predictLearner.classif.mxff = function(.learner, .model, .newdata, ...) {
+  x = data.matrix(.newdata)
+  array.layout = .model$learner$par.vals$array.layout
+  conv.layer1 = ifelse(is.null(.learner$par.vals$conv.layer1),
+    .learner$par.set$pars$conv.layer1$default, .learner$par.vals$conv.layer1)
+  if (conv.layer1) {
+    l = length(.learner$par.vals$conv.data.shape)
+    dims = switch(l,
+      c(.learner$par.vals$conv.data.shape, 1, 1, nrow(x)),
+      c(.learner$par.vals$conv.data.shape, 1, nrow(x)),
+      c(.learner$par.vals$conv.data.shape, nrow(x)),
+      .learner$par.vals$conv.data.shape)
+    x = array(aperm(x), dim = dims)
+    array.layout = "colmajor"
+  }
+  p = predict(.model$learner.model, X = x, array.layout = array.layout)
+  if (.learner$predict.type == "response") {
+    p = apply(p, 2, function(i) {
+      w = which.max(i)
+      return(ifelse(length(w > 0), w, NaN))
```

Comment:
I've commented this in the code. I think `NA` is more appropriate here, please tell me what you think.